data/src/main/java/com/microsoft/azure/kusto/data/BaseClient.java

```diff
@@ -37,79 +50,84 @@ public BaseClient(HttpClient httpClient) {
-            throw ExceptionUtils.createExceptionOnPost(e, request.getUrl(), "sync");
-        }
+    protected Mono<String> postAsync(HttpRequest request, long timeoutMs) {
+        return httpClient.send(request, getContextTimeout(timeoutMs))
+                .flatMap(response -> Utils.getResponseBody(response)
+                        .map(responseBody -> {
+                            switch (response.getStatusCode()) {
+                                case HttpStatus.OK:
+                                    return responseBody;
+                                case HttpStatus.TOO_MANY_REQS:
+                                    throw new ThrottleException(response.getRequest().getUrl().toString());
+                                default:
+                                    throw createExceptionFromResponse(
+                                            response.getRequest().getUrl().toString(), response, null, responseBody);
+                            }
+                        }).doFinally(ignore -> response.close()))
+                .onErrorMap(e -> {
+                    if (e instanceof DataServiceException) {
+                        return e;
+                    }
+                    return ExceptionUtils.createExceptionOnPost((Exception) e, request.getUrl(), "async");
+                });
     }
 
-    private String processResponseBody(HttpResponse response) throws DataServiceException {
-        String responseBody = Utils.isGzipResponse(response) ? Utils.gzipedInputToString(response.getBodyAsBinaryData().toStream())
-                : response.getBodyAsBinaryData().toString();
-
-        if (responseBody == null) {
-            return null;
-        }
-
-        switch (response.getStatusCode()) {
-            case HttpStatus.OK:
-                return responseBody;
-            case HttpStatus.TOO_MANY_REQS:
-                throw new ThrottleException(response.getRequest().getUrl().toString());
-            default:
-                throw createExceptionFromResponse(response.getRequest().getUrl().toString(), response, null, responseBody);
-        }
+    protected Mono<InputStream> postToStreamingOutputAsync(HttpRequest request, long timeoutMs,
+            int currentRedirectCounter, int maxRedirectCount) {
+        ResponseState state = new ResponseState();
+        return httpClient.send(request, getContextTimeout(timeoutMs))
+                .flatMap(httpResponse -> {
+                    state.setHttpResponse(httpResponse);
+                    int responseStatusCode = httpResponse.getStatusCode();
+                    if (responseStatusCode == HttpStatus.OK) {
+                        state.setReturnInputStream(true);
+                        return httpResponse.getBodyAsInputStream() // TODO: since we want to just close on EOF should we implement EofSensorInputStream
+                                                                   // ourselves and remove the remaining Apache dependency or not?
```

Comment:
If this is the last thing keeping the Apache dependency, I think we should implement it and remove the dependency.